### PR TITLE
improve the routing mechanism by augmenting the routing error type with ...

### PR DESCRIPTION
**Pull request by soenkehahn**
...some information and a sound monoid instance. this addresses #10

### CODE REVIEW

File: servant/src/Servant/API/Union.hs [hunk 1]
```diff
@@ -14,11 +14,16 @@ infixr 8 :<|>
...
 instance (HasServer a, HasServer b) => HasServer (a :<|> b) where
   type Server (a :<|> b) = Server a :<|> Server b
-  route Proxy (a :<|> b) request respond = do
-    route (Proxy :: Proxy a) a request $ \ mResponse ->
-      case mResponse of
-        Nothing -> route (Proxy :: Proxy b) b request respond
-        Just resp -> respond $ Just resp
+  route Proxy (a :<|> b) request respond =
+    route pa a request $ \ mResponse ->
+      case isMismatch mResponse of
```
Comment (soenkehahn) [inline servant/src/Servant/API/Union.hs]: @alpmestan: Why not `if-then-else`?

**Comment by alpmestan** [comment]
OK to merge this?

File: servant/src/Servant/API/Union.hs (hunk 1 above)
Comment (alpmestan) [inline servant/src/Servant/API/Union.hs]: Yeah we could use that if you prefer.

**Comment by soenkehahn** [comment]
Still looking through the diff.

File: servant/src/Servant/API/RQBody.hs [hunk 2]
```diff
@@ -26,7 +26,7 @@ instance (FromJSON a, HasServer sublayout)
   route Proxy subserver request respond = do
     mrqbody <- decode' <$> lazyRequestBody request
     case mrqbody of
-      Nothing -> respond Nothing
+      Nothing -> respond $ failWith NotFound
```
Comment (soenkehahn) [inline servant/src/Servant/API/RQBody.hs]: This means that the request body was either invalid JSON or couldn't be converted into the wanted type. This should yield a `400`.

Comment (alpmestan) [inline servant/src/Servant/API/RQBody.hs]: i'm going to fix this in a subsequent commit

**Pull request merged**